src/ray/object_manager/plasma/object_lifecycle_manager.h

```diff
@@ -0,0 +1,152 @@
+// It tracks object’s lifecycle states including reference_count and created/sealed.
+// It uses eviction_policy to garbage collect objects when running out of space.
+class ObjectLifecycleManager {
+ public:
+  ObjectLifecycleManager(IAllocator &allocator,
+                         ray::DeleteObjectCallback delete_object_callback);
+
+  /// Create a new object given object's info. Object creation might
+  /// fail if runs out of space; or an object with the same id exists.
+  ///
+  /// \param object_info Plasma object info.
+  /// \param source From where the object is created.
+  /// \param fallback_allocator Whether to allow fallback allocation.
+  /// \return
+  ///   - pointer to created object and PlasmaError::OK when succeeds.
+  ///   - nullptr and error message, including ObjectExists/OutOfMemory
+  std::pair<const LocalObject *, flatbuf::PlasmaError> CreateObject(
+      const ray::ObjectInfo &object_info, plasma::flatbuf::ObjectSource source,
+      bool fallback_allocator);
+
+  /// Get object by id.
+  /// \return
+  ///   - nullptr if such object doesn't exist.
+  ///   - otherwise, pointer to the object.
+  const LocalObject *GetObject(const ObjectID &object_id) const;
+
+  /// Seal created object by id.
+  ///
+  /// \param object_id Object ID of the object to be sealed.
+  /// \return
+  ///   - nulltpr if such object doesn't exist, or the object has already been sealed.
+  ///   - otherise, pointer to the sealed object.
+  const LocalObject *SealObject(const ObjectID &object_id);
+
+  /// Abort object creation by id. It deletes the object regardless of reference
+  /// counting.
+  ///
+  /// \param object_id Object ID of the object to be aborted.
+  /// \return
+  ///   - false if such object doesn't exist, or the object has already been sealed.
+  ///   - true if abort successfuly.
+  bool AbortObject(const ObjectID &object_id);
+
+  /// Delete a specific object by object_id. The object is delete immediately
+  /// if it's been sealed and reference counting is zero. Otherwise it will not
+  /// be deleted.
+  ///
+  /// \param object_id Object ID of the object to be deleted.
+  /// \return One of the following error codes:
+  ///  - PlasmaError::OK, if the object was delete successfully.
+  ///  - PlasmaError::ObjectNonexistent, if ths object isn't existed.
+  ///  - PlasmaError::ObjectNotsealed, if ths object is created but not sealed.
+  ///  - PlasmaError::ObjectInUse, if the object is in use; it will be deleted
+  ///  once it's no longer used (ref count becomes 0).
+  flatbuf::PlasmaError DeleteObject(const ObjectID &object_id);
+
+  /// Bump up the reference count of the object.
+  ///
+  /// \return true if object exists, false otherise.
+  bool AddReference(const ObjectID &object_id);
+
+  /// Decrese the reference count of the object. When reference count
+  /// drop to zero the object becomes evictable.
+  ///
+  /// \return true if object exists and reference count is greater than 0, false otherise.
```

Comment:
For this one, why don't we actually have a strong check here? I think having the correct reference count is very important, so making it less robust could be useful? Maybe we can just abort if the reference count == 0 when this method is called?

Reply:
as we discussed, let's only use RAY_CHECK for internal invariants. in this case the RAY_CHECK is not an invaraint internal to object_life_cycle_manager.